### ISSUE TITLE
We need FSharp.Compiler.SourceCodeServices.FSharpExprVisitor

### ISSUE THREAD

**Issue by Thorium** [fsharp/FSharp.Compiler.Service#938]
We need FSharp.Compiler.SourceCodeServices.FSharpExprVisitor

There is an ExpressionVisitor in core dotnet (System.Linq.Expressions.dll).
That is the way to do meta-programming in C#. It can be used to visit, traversing, examining, copying and transforming LINQ-expressions, which are the default Abstract Syntax Tree (AST) for C#.

In F# there is no one default AST.

- **F# Query expressions** are just a wrappers for LINQ-expressions. (You can use ExpressionVisitor like [this](http://fssnip.net/7Xg/title/F-LINQExpressiontree-visitor).)
- **Quotations. ASTs are Expr-class.** (There is no ExprVisior class, but a workaround is that you convert Expr to LINQ-expression, and use C# ExpressionVisitor. Ugly, but doable.)
- **FSharp.Compiler.SourceCodeServices.FSharpExpr** and this is the problematic one. Expression visitor cannot be done, because [type E](https://github.com/fsharp/FSharp.Compiler.Service/blob/314b89729ca0d8c00b1c555b48d07c03c2cb6d63/src/fsharp/symbols/Exprs.fs#L73) is not publicly visible and you cannot easily create a new FSharpExpr AST. The visitor should be written to this type.

By searching GitHub for e.g. [BasicPatterns.FastIntegerForLoop](https://github.com/search?p=1&q=BasicPatterns.FastIntegerForLoop&type=Code) we can see that there would really be use for this, as the workaround for people has been that they are writing their own yet-another-ASTs to traverse FSharpExpr to something and then processing that. Or even worse, not everyone is an AST-expert, so they try to collect some results directly, losing the tree, and then struggling to handle the results later.

So basically one public method should be added to Exprs.fs (and corresponding Exprs.fsi), which would do something like this:
```fsharp
let rec visitFSharpExpr memberCallHandler (e:FSharpExpr) =
    match e with
    | BasicPatterns.AddressOf(lvalueExpr) ->
        upcast FSharpExpr.NewAddressOf(visitFSharpExpr memberCallHandler lvalueExpr)
    | BasicPatterns.AddressSet(lvalueExpr, rvalueExpr) ->
        upcast FSharpExpr.NewAddressSet(visitFSharpExpr memberCallHandler lvalueExpr; visitFSharpExpr memberCallHandler rvalueExpr)
    | BasicPatterns.Application(funcExpr, typeArgs, argExprs) ->
       ///...
```
